internal/leadership/elector.go

```diff
@@ -62,206 +89,294 @@ type Elector struct {
+			case e.leadershipNotificationChan <- struct{}{}:
 			}
 		}
 
-		// notify all subscribers that we're the leader
-		e.notifySubscribers(true)
+		e.Logger.Info(e.Name+": Listening for leadership changes", "client_id", e.clientID, "topic", notifier.NotificationTopicLeadership)
 
-		err := e.keepLeadership(ctx, leadershipNotificationChan)
-		e.notifySubscribers(false)
-		if err != nil {
-			select {
-			case <-ctx.Done():
+		var err error
+		sub, err = notifier.ListenRetryLoop(ctx, &e.BaseService, e.notifier, notifier.NotificationTopicLeadership, handleNotification)
+		if err != nil { //nolint:staticcheck
+			// TODO(brandur): Propagate this after refactor.
+		}
+	}
+
+	go func() {
+		// This defer should come first so that it's last out, thereby avoiding
+		// races.
+		defer close(stopped)
+
+		e.Logger.InfoContext(ctx, e.Name+": Run loop started")
+		defer e.Logger.InfoContext(ctx, e.Name+": Run loop stopped")
+
+		if sub != nil {
+			defer sub.Unlisten(ctx)
+		}
+
+		for {
+			if err := e.attemptGainLeadershipLoop(ctx); err != nil {
+				// Function above only returns an error if context was cancelled
+				// or overall context is done.
+				if !errors.Is(err, context.Canceled) && ctx.Err() == nil {
+					panic(err)
+				}
 				return
-			default:
-				// TODO: backoff
-				e.logger.Error("error keeping leadership", "err", err)
-				continue
+			}
+
+			e.Logger.Info(e.Name+": Gained leadership", "client_id", e.clientID)
+			e.testSignals.GainedLeadership.Signal(struct{}{})
+
+			err := e.keepLeadershipLoop(ctx)
+			if err != nil {
+				if errors.Is(err, context.Canceled) {
+					return
+				}
+
+				if !errors.Is(err, errLostLeadership) {
+					e.Logger.Error(e.Name+": Error keeping leadership", "client_id", e.clientID, "err", err)
+				}
 			}
 		}
-	}
+	}()
+
+	return nil
 }
 
-func (e *Elector) gainLeadership(ctx context.Context, leadershipNotificationChan <-chan struct{}) bool {
+func (e *Elector) attemptGainLeadershipLoop(ctx context.Context) error {
+	var numErrors int
+
 	for {
-		success, err := e.attemptElect(ctx)
-		if err != nil && !errors.Is(err, context.Canceled) {
-			e.logger.Error("error attempting to elect", "err", err)
+		e.Logger.Info(e.Name+": Attempting to gain leadership", "client_id", e.clientID)
+
+		elected, err := attemptElectOrReelect(ctx, e.exec, false, &riverdriver.LeaderElectParams{
+			LeaderID: e.clientID,
+			Name:     e.instanceName,
+			TTL:      e.ttl,
+		})
+		if err != nil {
+			if errors.Is(err, context.Canceled) || ctx.Err() != nil {
+				return err
+			}
+
+			numErrors++
+			e.Logger.Error(e.Name+": Error attempting to elect", "client_id", e.clientID, "err", err, "num_errors", numErrors)
+			e.CancellableSleepExponentialBackoff(ctx, numErrors-1, baseservice.MaxAttemptsBeforeResetDefault)
+			continue
 		}
-		if success {
-			return true
+		if elected {
+			return nil
 		}
 
+		numErrors = 0
+
+		e.Logger.Info(e.Name+": Leadership bid was unsuccessful (not an error)", "client_id", e.clientID)
```

Comment:
`Info` level might be a bit noisy for this running every 5 seconds on every worker and it being a completely benign/expected scenario, maybe `Debug`?
```suggestion
		e.Logger.Debug(e.Name+": Leadership bid was unsuccessful (not an error)", "client_id", e.clientID)
```

Reply:
I'll say that because we probably don't expect _that_ many different River clients operating in one installation, so it's not like these will be getting endlessly spewed.

That said, now that the tests are all fixed up, I can live with debug, so changed.

Also, I changed everything to `InfoContext`/`DebugContext` which I'd meant to do but forgot.